File: src/main/java/io/lsdconsulting/lsd/distributed/interceptor/captor/http/ResponseCaptor.java
package io.lsdconsulting.lsd.distributed.interceptor.captor.http;

import feign.Response;
import io.lsdconsulting.lsd.distributed.access.model.InterceptedInteraction;
import io.lsdconsulting.lsd.distributed.interceptor.captor.convert.TypeConverter;
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive.HttpStatusDeriver;
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive.PathDeriver;
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive.SourceTargetDeriver;
import io.lsdconsulting.lsd.distributed.interceptor.captor.trace.TraceIdRetriever;
import io.lsdconsulting.lsd.distributed.interceptor.persistance.RepositoryService;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpRequest;
import org.springframework.http.client.ClientHttpResponse;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.Collection;
import java.util.Map;

import static io.lsdconsulting.lsd.distributed.access.model.InteractionType.RESPONSE;
import static org.apache.commons.lang3.StringUtils.EMPTY;

@RequiredArgsConstructor
@Slf4j
public class ResponseCaptor {

    private final RepositoryService repositoryService;
    private final SourceTargetDeriver sourceTargetDeriver;
    private final PathDeriver pathDeriver;
    private final TraceIdRetriever traceIdRetriever;
    private final HttpHeaderRetriever httpHeaderRetriever;
    private final HttpStatusDeriver httpStatusDeriver;
    private final String profile;

    @SneakyThrows
    public InterceptedInteraction captureResponseInteraction(final Response response, Long elapsedTime) {
        final var requestHeaders = httpHeaderRetriever.retrieve(response.request());
        final var responseHeaders = httpHeaderRetriever.retrieve(response);
        final String path = pathDeriver.derivePathFrom(response.request().url());
        final String target = sourceTargetDeriver.deriveTarget(requestHeaders, path);
        final String serviceName = sourceTargetDeriver.deriveServiceName(requestHeaders);
        final String traceId = traceIdRetriever.getTraceId(requestHeaders);
        final String httpStatus = httpStatusDeriver.derive(response.status());

        final InterceptedInteraction interceptedInteraction = buildInterceptedInteraction(target, path, traceId, elapsedTime, requestHeaders, responseHeaders, serviceName, TypeConverter.convert(response.body()), httpStatus);
        repositoryService.enqueue(interceptedInteraction);
        return interceptedInteraction;
    }

    public InterceptedInteraction captureResponseInteraction(final HttpRequest request, final ClientHttpResponse response, final String target, final String path, final String traceId, Long elapsedTime) throws IOException {
        final var requestHeaders = httpHeaderRetriever.retrieve(request);
        final var responseHeaders = httpHeaderRetriever.retrieve(response);
        final String serviceName = sourceTargetDeriver.deriveServiceName(requestHeaders);
        final String body = copyBodyToString(response);
        final String httpStatus = response.getStatusCode().toString();

        final InterceptedInteraction interceptedInteraction = buildInterceptedInteraction(target, path, traceId, elapsedTime, requestHeaders, responseHeaders, serviceName, body, httpStatus);
        repositoryService.enqueue(interceptedInteraction);
        return interceptedInteraction;
    }

    private InterceptedInteraction buildInterceptedInteraction(String target, String path, String traceId, Long elapsedTime, Map<String, Collection<String>> requestHeaders, Map<String, Collection<String>> responseHeaders, String serviceName, String body, String httpStatus) {
        return InterceptedInteraction.builder()
                .traceId(traceId)
                .body(body)
                .requestHeaders(requestHeaders)
                .responseHeaders(responseHeaders)
                .serviceName(serviceName)
                .target(target)
                .path(path)
                .httpStatus(httpStatus)
                .interactionType(RESPONSE)
                .profile(profile)
                .elapsedTime(elapsedTime)
                .createdAt(ZonedDateTime.now(ZoneId.of("UTC")))
                .build();
    }

    private String copyBodyToString(final ClientHttpResponse response) throws IOException {
        if (response.getHeaders().getContentLength() == 0) {
            return EMPTY;
        }
        final ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        final InputStream inputStream = response.getBody();
        inputStream.transferTo(outputStream);
        return outputStream.toString();
    }
}

File: src/main/kotlin/io/lsdconsulting/lsd/distributed/interceptor/captor/http/ResponseCaptor.kt
package io.lsdconsulting.lsd.distributed.interceptor.captor.http

import feign.Response
import io.lsdconsulting.lsd.distributed.access.model.InteractionType
import io.lsdconsulting.lsd.distributed.access.model.InterceptedInteraction
import io.lsdconsulting.lsd.distributed.interceptor.captor.convert.stringify
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive.SourceTargetDeriver
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive.toHttpStatus
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive.toPath
import io.lsdconsulting.lsd.distributed.interceptor.captor.trace.TraceIdRetriever
import io.lsdconsulting.lsd.distributed.interceptor.persistance.RepositoryService
import org.apache.commons.lang3.StringUtils
import org.springframework.http.HttpRequest
import org.springframework.http.client.ClientHttpResponse
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.time.ZoneId
import java.time.ZonedDateTime

class ResponseCaptor(
    private val repositoryService: RepositoryService,
    private val sourceTargetDeriver: SourceTargetDeriver,
    private val traceIdRetriever: TraceIdRetriever,
    private val httpHeaderRetriever: HttpHeaderRetriever,
    private val profile: String,
) {
    fun captureResponseInteraction(response: Response, elapsedTime: Long): InterceptedInteraction {
        val requestHeaders = httpHeaderRetriever.retrieve(response.request())
        val responseHeaders = httpHeaderRetriever.retrieve(response)
        val path = response.request().url().toPath()
        val target = sourceTargetDeriver.deriveTarget(requestHeaders, path)
        val serviceName = sourceTargetDeriver.deriveServiceName(requestHeaders)
        val traceId = traceIdRetriever.getTraceId(requestHeaders)
        val httpStatus = response.status().toHttpStatus()
        val interceptedInteraction = buildInterceptedInteraction(
            target,
            path,
            traceId,
            elapsedTime,
            requestHeaders,
            responseHeaders,
            serviceName,
            response.body()?.stringify(),
            httpStatus
        )
        repositoryService.enqueue(interceptedInteraction)
        return interceptedInteraction
    }

    @Throws(IOException::class)
    fun captureResponseInteraction(
        request: HttpRequest,
        response: ClientHttpResponse,
        target: String,
        path: String,
        traceId: String,
        elapsedTime: Long
    ): InterceptedInteraction {
        val requestHeaders = httpHeaderRetriever.retrieve(request)
        val responseHeaders = httpHeaderRetriever.retrieve(response)
        val serviceName = sourceTargetDeriver.deriveServiceName(requestHeaders)
        val body = copyBodyToString(response)
        val httpStatus = response.statusCode.toString()
        val interceptedInteraction = buildInterceptedInteraction(
            target,
            path,
            traceId,
            elapsedTime,
            requestHeaders,
            responseHeaders,
            serviceName,
            body,
            httpStatus
        )
        repositoryService.enqueue(interceptedInteraction)
        return interceptedInteraction
    }

    private fun buildInterceptedInteraction(
        target: String,
        path: String,
        traceId: String,
        elapsedTime: Long,
        requestHeaders: Map<String, Collection<String>>,
        responseHeaders: Map<String, Collection<String>>,
        serviceName: String,
        body: String?,
        httpStatus: String?
    ) = InterceptedInteraction(
        traceId = traceId,
        body = body,
        requestHeaders = requestHeaders,
        responseHeaders = responseHeaders,
        serviceName = serviceName,
        target = target,
        path = path,
        httpStatus = httpStatus,
        httpMethod = null, interactionType = InteractionType.RESPONSE,
        profile = profile,
        elapsedTime = elapsedTime,
        createdAt = ZonedDateTime.now(ZoneId.of("UTC"))
    )

    @Throws(IOException::class)
    private fun copyBodyToString(response: ClientHttpResponse): String {
        if (response.headers.contentLength == 0L) {
            return StringUtils.EMPTY
        }
        val outputStream = ByteArrayOutputStream()
        val inputStream = response.body
        inputStream.transferTo(outputStream)
        return outputStream.toString()
    }
}